### PR TITLE
as-tree: init at 0.12.0

### PR DESCRIPTION
**Pull request by jshholland**
###### Motivation for this change

Add new package as-tree.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/as-tree/default.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  src = fetchFromGitHub {
+    owner = "jez";
+    repo = pname;
+    rev = version;
+    sha256 = "0c0g32pkyhyvqpgvzlw9244c80npq6s8mxy3may7q4qyd7hi3dz5";
+  };
+
+  cargoSha256 = "0g3x76l251g82jrd194whdsfxzlg45bhr1y6bjjfc79viz9mrr87";
+  cargoPatches = [ ./cargo-lock.patch ];
```
Comment (syberant) [inline pkgs/tools/misc/as-tree/default.nix:15]: As far as I understand the purpose of this patch is to update the erroneous version of as-tree in Cargo.lock but it also updates the (minor) versions of other crates.
Might I suggest updating the `rev` in `fetchFromGitHub` to use the [upstream fix](https://github.com/jez/as-tree/commit/c82bab213ad11fa52a7827146b506f44aeef4c25) instead? I (personally) think it's cleaner.

**Review by syberant — COMMENTED** [review]
I've also tested building and running the binary on NixOS and haven't encountered any errors.

File: pkgs/tools/misc/as-tree/default.nix (hunk 1 above)
Comment (jshholland) [reply pkgs/tools/misc/as-tree/default.nix:15]: Possibly: I saw that commit also, but I wanted to stick to the tagged release commit. Ideally, I suppose, upstream would have a 0.12.1 release to fix this. I'm new to packaging Rust applications for Nix. Perhaps the ideal middle ground is to change the patch so it only fixes the wrong version of as-tree, and leaves the other crates as-is?

Comment (syberant) [reply pkgs/tools/misc/as-tree/default.nix:15]: Fine with me. Maybe also leave a small note above cargoPatches to indicate why the patch exists and to remove it when a new version releases.

File: pkgs/tools/misc/as-tree/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  verifyCargoDeps = true;
```
Comment (syberant) [inline pkgs/tools/misc/as-tree/default.nix:16]:
```suggestion
```
Someone pointed out to me today that `verifyCargoDeps` [has been removed](https://github.com/NixOS/nixpkgs/issues/79975).

Comment (jshholland) [reply pkgs/tools/misc/as-tree/default.nix]: Done

File: pkgs/tools/misc/as-tree/default.nix (hunk 1 above)
Comment (jshholland) [reply pkgs/tools/misc/as-tree/default.nix:18]: I have added an explanatory comment and reduced the patch to only update the version of as-tree itself

**Comment by jshholland** [comment]
Rebased and added changes suggested by @syberant, thanks for the review and apologies it's taken so long to fix it up!

**Review by syberant — APPROVED** [review]
Works fine when testing with `nixpkgs-review pr 94074`.

**Comment by eyJhb** [comment]
Works fine for me as well :)

**Review by stigtsp — APPROVED** [review]
LGTM

Built and tested OK on `linux x64`

Result of `nixpkgs-review pr 94074` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>as-tree</li>
  </ul>
</details>

**Comment by jshholland** [comment]
Thanks! My first package into nixpkgs :)

**Comment by stigtsp** [comment]
> Thanks! My first package into nixpkgs :)

Thank you for the contribution, and welcome to the NixOS project :tada:

**Pull request merged**